pkg/ebpf/netlink/attribute_scanner.go

```diff
@@ -0,0 +1,192 @@
+var attrTypeMask uint16 = 0x3fff
+
+// errors
+var (
+	errInvalidAttribute  = errors.New("invalid attribute; length too short or too large")
+	errMessageTooShort   = errors.New("netlink message is too short")
+	errMissingNestedAttr = errors.New("netlink message is missing nested attribute")
+)
+
+// Based on https://github.com/mdlayher/netlink/blob/master/attribute.go
+// The main optimizations here are:
+// * We don't allocate a slice for the field data when we're scanning the attributes;
+// * The AttributeScanner itself can be reused when scanning nested fields and decoding different messages;
+type AttributeScanner struct {
+	level  int
+	frames []*NestedFrame
+}
+
+// A NestedFrame encapsulates the decoding information of a certain nesting level
+type NestedFrame struct {
+	a netlink.Attribute
+
+	// The slice of input bytes and its iterator index.
+	b []byte
+	i int
+
+	// Any error encountered while decoding attributes.
+	err error
+}
+
+// NewAttributeScanner returns a new instance of AttributeScanner
+func NewAttributeScanner() *AttributeScanner {
+	scanner := &AttributeScanner{
+		frames: make([]*NestedFrame, 3),
+	}
+
+	for i := range scanner.frames {
+		scanner.frames[i] = &NestedFrame{}
+	}
+
+	return scanner
+}
+
+// Next advances the decoder to the next netlink attribute.  It returns false
+// when no more attributes are present, or an error was encountered.
+func (s *AttributeScanner) Next() bool {
+	f := s.currentFrame()
+
+	if f.err != nil {
+		// Hit an error, stop iteration.
+		return false
+	}
+
+	// Exit if array pointer is at or beyond the end of the slice.
+	if f.i >= len(f.b) {
+		return false
+	}
+
+	if err := f.unmarshal(); err != nil {
+		f.err = err
+		return false
+	}
+
+	// Advance the pointer by at least one header's length.
+	if int(f.a.Length) < nlaHeaderLen {
+		f.i += nlaHeaderLen
+	} else {
+		f.i += nlaAlign(int(f.a.Length))
+	}
+
+	return true
+}
+
+// Type returns the Attribute.Type field of the current netlink attribute
+// pointed to by the decoder.
+//
+// Type masks off the high bits of the netlink attribute type which may contain
+// the Nested and NetByteOrder flags. These can be obtained by calling TypeFlags.
+func (s *AttributeScanner) Type() uint16 {
+	// Mask off any flags stored in the high bits.
+	return s.currentFrame().a.Type & attrTypeMask
+}
+
+// Err returns the first error encountered by the scanner.
+func (s *AttributeScanner) Err() error {
+	return s.currentFrame().err
+}
+
+// Bytes returns the raw bytes of the current Attribute's data.
+func (s *AttributeScanner) Bytes() []byte {
+	return s.currentFrame().a.Data
+}
+
+// Nested executes the given function within a new NestedFrame
+func (s *AttributeScanner) Nested(fn func() error) {
+	if len(s.frames) <= s.level {
+		s.frames = append(s.frames, &NestedFrame{})
+	}
+
+	prev := s.currentFrame()
+
+	// Create new frame
+	s.level++
+	current := s.currentFrame()
+	current.b = prev.a.Data
```

Comment:
Should this be only the part prev.a.Data that is part of the nested part of the message?

I think:  `prev.Data[prev.i + nlaHeaderLen:prev.i + nlaHeaderLength + prev.a.Length]`

Reply:
Not really. the Data field from the Attr has only the "payload", and not the header data:
https://github.com/DataDog/datadog-agent/blob/pedro/conntrack-refactor/pkg/ebpf/netlink/attribute_scanner.go#L198